### PR TITLE
Add PHP8 Unit Testing

### LINKED ISSUE #7425: Include PHP/WC versions in our testing pipeline ⚙️
## Is your feature request related to a problem? Please describe.
Currently, we run our automated testing jobs on PHP 7.4, which is not ideal since we support multiple PHP versions, and 7.4 will discontinue security updates in late November.

## Describe the solution you'd like
We already have Unit and E2E testing jobs in place on our pipeline. I consider low-hanging fruit (not much low but still) duplicating those jobs and having them running against our supported PHP (and possibly WC) versions. We might consider alternative solutions as long as they achieve similar results while evolving smaller efforts.
This is supposed to be an amount of work that can be tackled during a cooldown.

The WooCommerce repo already does something similar:
image

**EDIT:** After gathering some thoughts at Rubik's team meeting, it was proposed that a good course of action would be:
- Running PHP Unit Tests against 7.4 & 8.0 (low minute cost)
- ~~Change E2E to run against PHP 8.0 due to 7.4 security support being dropped soon~~
- Outside the scope of this task, we might pursue additional testing on nightlies/scheduled runs against multiple combinations of PHP/WC/WP

## Additional context
This will not be a definitive solution, as upstream efforts are in place to [migrate E2E testing to Playwright](https://masamunep2.wordpress.com/2022/08/26/testing-infrastructure-proposal/).
The goal here is to deploy with more confidence and avoid any related tech debt to be dealt with in the future (eg: introducing incompatible code/bugs)

[Slack thread for reference.](p1665745482034199-slack-C02UBB1EPEF)

### PR DESCRIPTION
**Pull request by wavvves**
Fixes #7425

The purpose of this PR is to duplicate PHP Unit tests to run against PHP 7.4 and 8.0.

Unit testing runs under wp-env, and PHP runtime is configured under `.wp-env.json`. This PR uses a [matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) for splitting the job and generates a `.wp-env.override.json`, which effectively sets the [PHP version used on wp-env instances](https://github.com/WordPress/gutenberg/pull/25268/files).

I encountered a problem where previous cached wp-env data would interfere with the test output, so data is deleted between regular tests, and WC dev version test.

~`@wordpress/env` was also updated to v5~ and it now uses PHPUnit 9.x (our dev dependency binary) as it was an opportunity to get rid of that tech debt.
Some tests were required to be updated to work with that.

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### Testing

1. Inspect Github Jobs for this PR
2. Ensure Unit Testing now reports **passing tests** for PHP 7.4 and PHP 8.0

* [x] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> n/a

### CODE REVIEW

File: src/Migration.php [hunk 1]
```diff
@@ -52,4 +50,15 @@ public function run_migrations() {
...
 	}
+
+	/**
+	 * Sets the $db_upgrades array value.
+	 *
+	 * @param array $db_upgrades Array of migrations to run.
+	 *
+	 * @return void
+	 */
+	public function set_db_upgrades( array $db_upgrades ) {
```
Comment (gigitux) [inline src/Migration.php]: Can you explain why did you add this function?

File: tests/php/StoreApi/Routes/ProductAttributes.php [hunk 2]
```diff
@@ -53,7 +53,7 @@ public function test_get_items() {
 		$data     = $response->get_data();
 
 		$this->assertEquals( 200, $response->get_status() );
-		$this->assertEquals( 2, count( $data ) );
+		$this->assertEquals( count( wc_get_attribute_taxonomies() ), count( $data ) );
```
Comment (gigitux) [inline tests/php/StoreApi/Routes/ProductAttributes.php:56]: Is it necessary? Isn't it better to hardcode the value?

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks a lot for this effort! 🙇 
I have some questions regarding your code changes. Also, it seems that this change breaks completely E2E tests.

When you will merge this PR, remember to update [the branch protection rules](https://github.com/orgs/community/discussions/26698).

image

File: tests/php/StoreApi/Routes/ProductAttributes.php (hunk 2 above)
Comment (wavvves) [reply tests/php/StoreApi/Routes/ProductAttributes.php:56]: I was getting mismatched results between dev and CI (most probably because wp-env isn't stateless between steps or even jobs it seems 🙄)

Comparing the result coming from the API with wc_get_attribute_taxonomies() seems more robust since those should walk hand to hand. If we hardcode 2, but the count from taxonomies returns 6, and the test passes, would you say we have good code? Or if both return the same value but a wrong value, would you say it is our code's fault? 🤔

File: src/Migration.php (hunk 1 above)
Comment (wavvves) [reply src/Migration.php]: I removed the reflection class because I was getting weird issues (a mix of PHP incompatibility and ci cache after all) and replaced that with a setter to keep code simple, but forgot to change it back.

Nice catch, I'm reverting that 👍🏼

File: tests/php/StoreApi/Routes/CartItems.php [hunk 3]
```diff
@@ -15,9 +15,9 @@
 class CartItems extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/CartItems.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/CartExtensions.php [hunk 4]
```diff
@@ -14,9 +14,9 @@
 class CartExtensions extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/CartExtensions.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/CartCoupons.php [hunk 5]
```diff
@@ -15,9 +15,9 @@
 class CartCoupons extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/CartCoupons.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/Cart.php [hunk 6]
```diff
@@ -17,9 +17,9 @@
 class Cart extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/Cart.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/Batch.php [hunk 7]
```diff
@@ -14,9 +14,9 @@
 class Batch extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/Batch.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Formatters/TestMoneyFormatter.php [hunk 8]
```diff
@@ -15,9 +15,9 @@ class TestMoneyFormatter extends \WP_UnitTestCase {
 	private $mock_formatter;
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Formatters/TestMoneyFormatter.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Formatters/TestHtmlFormatter.php [hunk 9]
```diff
@@ -15,9 +15,9 @@ class TestHtmlFormatter extends \WP_UnitTestCase {
 	private $mock_formatter;
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Formatters/TestHtmlFormatter.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Formatters/TestFormatters.php [hunk 10]
```diff
@@ -19,9 +19,9 @@ class TestFormatters extends \WP_UnitTestCase {
 	private $mock_formatters;
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Formatters/TestFormatters.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Formatters/TestCurrencyFormatter.php [hunk 11]
```diff
@@ -15,9 +15,9 @@ class TestCurrencyFormatter extends \WP_UnitTestCase {
 	private $mock_formatter;
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Formatters/TestCurrencyFormatter.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/ExtendSchemaTests.php [hunk 12]
```diff
@@ -29,7 +29,7 @@ class ExtendSchemaTests extends TestCase {
 	private $dummy;
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/ExtendSchemaTests.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/Checkout.php [hunk 13]
```diff
@@ -22,7 +22,7 @@
  */
 class Checkout extends MockeryTestCase {
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/Checkout.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/ProductAttributes.php [hunk 14]
```diff
@@ -15,9 +15,9 @@
 class ProductAttributes extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/ProductAttributes.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/ProductCollectionData.php [hunk 15]
```diff
@@ -15,9 +15,9 @@
 class ProductCollectionData extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/ProductCollectionData.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Routes/Products.php [hunk 16]
```diff
@@ -15,9 +15,9 @@
 class Products extends ControllerTestCase {
 
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Routes/Products.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

File: tests/php/StoreApi/Utilities/ProductQueryFilters.php [hunk 17]
```diff
@@ -7,9 +7,9 @@
 
 class ProductQueryFiltersTest extends TestCase {
 	/**
-	 * Setup test products data. Called before every test.
+	 * Setup test products' data. Called before every test.
```
Comment (mikejolley) [inline tests/php/StoreApi/Utilities/ProductQueryFilters.php]:
```suggestion
	 * Setup test product data. Called before every test.
```

**Review by mikejolley — COMMENTED** [review]
Changes seem okay to me, so assuming tests pass ship it.

Are the e2e fails unrelated?

**Comment by wavvves** [comment]
> Changes seem okay to me, so assuming tests pass ship it.
> 
> 
> 
> Are the e2e fails unrelated?

No, I believe they are related with the @wordpress/env upgrade. I'm going to downgrade it, and see if I can find another way to fix the problems it was causing.

**Comment by wavvves** [comment]
@gigitux @mikejolley E2E tests now "fail successfully" (🙄) after downgrading @wordpress/env, is this PR good to go?

I found some issues that will spawn separate issues.

**Review by gigitux — APPROVED** [review]
Except for @mikejolley's comments, LGTM! 
Thanks for taking care of this 🙇

**Review by mikejolley — COMMENTED** [review]
If the fails match trunk go ahead @wavvves ? I know the Gutenberg ones are failing, but there are also some fails from latest which aren't clear if they should be failing or not, or if related to these changes.

**Comment by wavvves** [comment]
> If the fails match trunk go ahead @wavvves ? I know the Gutenberg ones are failing, but there are also some fails from latest which aren't clear if they should be failing or not, or if related to these changes.

I compared these with other branches ([example](https://github.com/woocommerce/woocommerce-blocks/actions/runs/3428968798/jobs/5713958515)) and failing seems consistent.

I'm going to merge this. My hint about wp-env state being the culprit for failing E2E gets stronger everyday.
